### PR TITLE
add AzureKeyCredentialPolicy.cs into project file

### PR DESCRIPTION
**Pull request by chunyu3**
# Description
Fix https://github.com/Azure/autorest.csharp/issues/3773
Fix https://github.com/Azure/autorest.csharp/issues/3569
- Add AzureKeyCredentialPolicy.cs into project file(.csproj) when the sdk need azure key authentication.

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: test/CadlRanchProjects/authentication/api-key/src/Authentication.ApiKey.csproj [hunk 1]
```diff
@@ -10,6 +10,7 @@
 
   <ItemGroup>
     <Compile Include="$(AzureCoreSharedSources)AzureResourceProviderNamespaceAttribute.cs" LinkBase="Shared/Core" />
+    <Compile Include="$(AzureCoreSharedSources)AzureKeyCredentialPolicy.cs" LinkBase="Shared/Core" />
```
Comment (lirenhe) [inline test/CadlRanchProjects/authentication/api-key/src/Authentication.ApiKey.csproj:13]: what is the impact if this line was not added?  Seems we didn't get any problem of the compile error.

Comment (chunyu3) [reply test/CadlRanchProjects/authentication/api-key/src/Authentication.ApiKey.csproj:13]: Yes. It used to be added in `Directory.Build.targets` for autorest.csharp local test projects. so those test projects can compile successfully in autorest.csharp. But for those outside projects, without this, build will fail if the project has azure key authentication. To resolve this compile failure, we will add this to each sdk project file, and remove it from `Directory.Build.targets`.

**Comment by pshao25** [comment]
I thought the solution is supposed to be: only when azure key credential is used, we add it to project file.

This solution looks a workaround to me.

**Comment by chunyu3** [comment]
> only when azure key credential is used, we add it to project file.
> 
> This solution looks a workaround to me.

Yes. @pshao25 we only add `AzurekeyCredentialPolicy` to project file when azure key credential is used. See  in the code we will check the actual authentication used, then decide to add this or not.

File: src/AutoRest.CSharp/Common/AutoRest/Plugins/NewProjectScaffolding.cs [hunk 2]
```diff
@@ -16,14 +16,16 @@ internal class NewProjectScaffolding
         private string _testDirectory;
         private string _serviceDirectory;
         private bool _isAzureSdk;
+        private bool _NeedAzureKeyAuth;
```
Comment (pshao25) [inline src/AutoRest.CSharp/Common/AutoRest/Plugins/NewProjectScaffolding.cs:19]:
```suggestion
        private bool _needAzureKeyAuth;
```

**Review by pshao25 — APPROVED** [review]

File: src/AutoRest.CSharp/Common/AutoRest/Plugins/NewProjectScaffolding.cs [hunk 3]
```diff
@@ -292,7 +294,13 @@ private string GetSrcCsproj()
   </PropertyGroup>
 
   <ItemGroup>
-    <Compile Include=""$(AzureCoreSharedSources)AzureResourceProviderNamespaceAttribute.cs"" LinkBase=""Shared/Core"" />
+    <Compile Include=""$(AzureCoreSharedSources)AzureResourceProviderNamespaceAttribute.cs"" LinkBase=""Shared/Core"" />";
+            if (_needAzureKeyAuth)
```
Comment (archerzz) [inline src/AutoRest.CSharp/Common/AutoRest/Plugins/NewProjectScaffolding.cs:298]: You actually can define `srcBrandedCsprojContent` as `@$"xxxx"`, e.g. multiple line formattable string. So there the codes can be like:
```suggestion
           { (_needAzureKeyAuth ? @"....." : @"....")}
```

**Review by archerzz — COMMENTED** [review]
Can we have a test case which requires `AzureKeyCredentialPolicy.cs` to compile?

File: src/AutoRest.CSharp/Common/AutoRest/Plugins/NewProjectScaffolding.cs (hunk 3 above)
Comment (chunyu3) [reply src/AutoRest.CSharp/Common/AutoRest/Plugins/NewProjectScaffolding.cs:298]: hello @archerzz  I have thought about this. but it cannot work for some encoding issue. I remember. So we need to maintain as it is.

**Comment by chunyu3** [comment]
> Can we have a test case which requires `AzureKeyCredentialPolicy.cs` to compile?

Actually there are some projects in autorest.csharp which need `AzureKeyCredentialPolicy.cs`, so in this PR, we update the project files for them. No need to add more.

**Review by ArcturusZhang — APPROVED** [review]

**Review by archerzz — APPROVED** [review]

**Review by lirenhe — COMMENTED** [review]
Could you provide the link of the Regen PR?  Thanks

**Review by lirenhe — APPROVED** [review]

**Pull request merged**